core/primitives/src/shard_layout.rs

```diff
@@ -90,11 +108,62 @@ fn is_top_level_account(top_account: &AccountId, account: &AccountId) -> bool {
         Some(rest) => rest.is_empty() || rest.ends_with("."),
     }
 }
+
+#[derive(Hash, Clone, Debug, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ShardUId {
+    pub version: ShardVersion,
+    pub shard_id: u32,
+}
+
+impl ShardUId {
+    pub fn default() -> Self {
+        Self { version: 0, shard_id: 0 }
+    }
+    pub fn to_vec(&self) -> [u8; 8] {
+        let mut res = [0; 8];
+        res[0..4].copy_from_slice(&u32::to_le_bytes(self.version));
+        res[4..].copy_from_slice(&u32::to_le_bytes(self.shard_id));
+        res
+    }
+
+    pub fn from_vec(bytes: &[u8]) -> Result<Self, Box<dyn std::error::Error>> {
+        if bytes.len() != 8 {
+            return Err("incorrect length for ShardUId".into());
+        }
+        let version = u32::from_le_bytes(bytes[0..4].try_into().unwrap());
+        let shard_id = u32::from_le_bytes(bytes[4..8].try_into().unwrap());
+        Ok(Self { version, shard_id })
+    }
+}
+
+pub fn get_block_shard_uid(block_hash: &CryptoHash, shard_uid: ShardUId) -> Vec<u8> {
+    let mut res = Vec::with_capacity(40);
+    res.extend_from_slice(block_hash.as_ref());
+    res.extend_from_slice(&shard_uid.to_vec());
+    res
+}
+
+#[allow(unused)]
+pub fn get_block_shard_uid_rev(
+    key: &[u8],
+) -> Result<(CryptoHash, ShardUId), Box<dyn std::error::Error>> {
```

Comment:
We generally avoid generic error types such as `std::error::Error` because it creates trouble when we have to convert error types between each other. It is better to create a specific error type

Reply:
The signature of `get_block_shard_id_rev` also uses `Box<dyn std::error::Error>` and this function will only be used to replace `get_block_shard_id_rev` in selected places, so I think it should be fine to just use `std::error:Error`